File: updater/src/main/java/de/darkatra/patcher/updater/service/model/Packet.java
package de.darkatra.patcher.updater.service.model;

import lombok.Data;
import lombok.experimental.Accessors;

import java.nio.file.Paths;
import java.time.Instant;

@Data
@Accessors(chain = true)
public class Packet implements ContextAware {
	private String src;
	private String dest;
	private Long packetSize;
	private Instant dateTime;
	private String checksum;
	private boolean backupExisting;
	private Compression compression;

	@Override
	public void applyContext(final Context context) {
		context.forEach((key, value) -> {
			src = src.replace(context.getPrefix() + key + context.getSuffix(), value);
			dest = dest.replace(context.getPrefix() + key + context.getSuffix(), value);
		});
		dest = Paths.get(dest).normalize().toString();
	}
}

File: src/jvmMain/kotlin/de/darkatra/bfme2/patch/Packet.kt
package de.darkatra.bfme2.patch

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import java.nio.file.Path
import java.time.Instant

data class Packet(
    var src: String,
    var dest: String,
    @JsonProperty("packetSize")
    val size: Long,
    val compressedSize: Long,
    val dateTime: Instant,
    val checksum: String,
    val backupExisting: Boolean,
    @JsonDeserialize(converter = CompressionDeserializer::class)
    val compression: Compression
) : ContextAware {

    /**
     * Applies the given [Context] to the [Packet], replacing placeholders with their actual value.
     * Ideally this operation should only be performed once per [Packet].
     *
     * @param context the [Context] containing values for placeholders
     */
    override fun applyContext(context: Context) {
        context.forEach { key, value ->
            src = src.replace("${Context.PREFIX}$key${Context.SUFFIX}", value)
            dest = Path.of(dest.replace("${Context.PREFIX}$key${Context.SUFFIX}", value)).normalize().toString()
        }
    }
}
